src/mnode/mnode-pastel.cpp

```diff
@@ -1571,3 +1571,172 @@ bool CTakeDownTicket::FindTicketInDb(const std::string& key, CTakeDownTicket& ti
+            throw std::runtime_error(strprintf("%s ticket is invalid. Already changed in last 24 hours. PastelID - [%s]", GetTicketDescription(TicketID::Username), pastelID));
+        }
+    }
+
+    // E. Check if ticket fee is valid
+    if (!foundTicketBySecondaryKey) {
+        if (fee != masterNodeCtrl.MasternodeUsernameFirstChangeFee) {
+            throw std::runtime_error(strprintf("%s ticket's fee is invalid. PastelID - [%s], invalid fee - [%" PRId64 "], expected fee - [%" PRId64 "]",
+                                               GetTicketDescription(TicketID::Username), pastelID, fee, masterNodeCtrl.MasternodeUsernameFirstChangeFee));
+        }
+    } else {
+        if (fee != masterNodeCtrl.MasternodeUsernameChangeAgainFee) {
+            throw std::runtime_error(strprintf("%s ticket's fee is invalid. PastelID - [%s], invalid fee - [%" PRId64 "], expected fee - [%" PRId64 "]",
+                                               GetTicketDescription(TicketID::Username), pastelID, fee, masterNodeCtrl.MasternodeUsernameChangeAgainFee));
+        }
+    }
+    
+    return true;
+}
+
+CChangeUsernameTicket CChangeUsernameTicket::Create(std::string _pastelID, std::string _username, const SecureString& strKeyPass)
+{
+    CChangeUsernameTicket ticket(std::move(_pastelID), std::move(_username));
+
+    // Check if PastelID already have a username on the blockchain. 
+    if (!masterNodeCtrl.masternodeTickets.FindTicketBySecondaryKey(ticket)) {
+        // IF PastelID has no Username yet, the fee is 100 PSL
+        ticket.fee = masterNodeCtrl.MasternodeUsernameFirstChangeFee;
+    } else {
+        // IF PastelID changed Username before, fee should be 5000
+        ticket.fee = masterNodeCtrl.MasternodeUsernameChangeAgainFee;
+    }
+
+    ticket.GenerateTimestamp();
+    
+    std::string strTicket = ticket.ToStr();
+    ticket.signature = CPastelID::Sign(reinterpret_cast<const unsigned char*>(strTicket.c_str()), strTicket.size(), ticket.pastelID, strKeyPass);
+
+    return ticket;
+}
+
+bool CChangeUsernameTicket::FindTicketInDb(const std::string& key, CChangeUsernameTicket& ticket)
+{
+    ticket.username = key;
+    return masterNodeCtrl.masternodeTickets.FindTicket(ticket);
+}
+
+bool CChangeUsernameTicket::isUsernameBad(const std::string& username, std::string& error)
+{
+    std::string upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+    std::string lower = "abcdefghijklmnopqrstuvwxyz";
+    std::string digits = "0123456789";
+
+    // Check if has only <4, or has more than 12 characters
+    if ((username.size() < 4) || (username.size() > 12)) {
+        error = "Invalid size of username, the size should have at least 4 characters, and at most 12 characters";
+        return true;
+    }
+
+    // Check if doesn't start with letters.
+    if ( (upper.find(username.front()) == std::string::npos) && (lower.find(username.front()) == std::string::npos ) ) {
```

Comment:
yes I was going to use isalpha before. But then I realize that the isalpha result will depend on the locale. For example, it returns true for `unsigned char c = '\xdf';` for ISO-8859-1 locale in this [example](https://en.cppreference.com/w/cpp/string/byte/isalpha). 
So I moved to this approach to make it clearer and prevent dealing with locale.

Reply:
can you add then functions to str_utils.h like these:
inline bool isalphaex(const char c) noexcept { return ((c >= 'A') && (c <= 'Z')) || ((c >= 'a') && (c <= 'z')); }
inline bool isdigitex(const char c) noexcept { return (c >= '0') && (c <= '9'); }
and use instead of string search. I wrote small test program to compare two approaches and using isalphaex etc functions is ~8 times faster.

Reply:
fixed